src/core/manager.c

```diff
@@ -1828,17 +1829,19 @@ static int manager_dispatch_signal_fd(sd_event_source *source, int fd, uint32_t
         }
 
         for (;;) {
-                n = read(m->signal_fd, &sfsi, sizeof(sfsi));
-                if (n != sizeof(sfsi)) {
-
-                        if (n >= 0)
-                                return -EIO;
-
-                        if (errno == EINTR || errno == EAGAIN)
+                ttl = 0;
+                do {
+                        n = read(m->signal_fd, buf + ttl, length - ttl);
+                        if (n < 0 && errno == EAGAIN) {
                                 break;
-
-                        return -errno;
-                }
+                        } else if (n < 0 && errno != EINTR) {
+                                log_error_errno(errno, "Failed to read bytes from signal file descriptor (%d): %m", errno); return -errno;
+                        } else if (n == 0) {
+                            log_error("Signal file descriptor received client disconnect, with %ld bytes discarded.", ttl); return -EIO; }
```

Comment:
> The client is the remote end of the socket - the kernel pushing wrapped signal detail messages on to the signalfd(2) socket file descriptor.

Hm, where is the socket?
https://github.com/torvalds/linux/blob/master/fs/signalfd.c